### PR TITLE
Add ezquo module

### PR DESCRIPTION
**Pull request by drusepth**
Probably good to note: this module doesn't work. Scrollback searching isn't getting the correct lines.

Intent: this module allows you to specify a nick and a search limit (e.g. `ek 5`) and it'll split out the last 5 lines that person has said, already reverse-indexed to assist in out-of-context !quo's.

Example use:
```
01:18 < drusepth> !ezquo drusepth 3
01:18 < ^> (3) <drusepth> I'm so great
01:18 < ^> (2) <drusepth> I built this broken feature
01:18 < ^> (1) <drusepth> Okay time to sleep
```

Would be nice if it could also support regex/keyword searches, so you can quickly find the offset of the last N times @hfukada talked about kpop.

e.g.
```
01:18 < drusepth> !ezquo suroi 2 /kpop/
01:18 < ^> (98) <suroi> kpop is the bomb diggity
01:18 < ^> (32) <suroi> this is a real quote about kpop
```

Does not re-print-out #noquo lines.

Assigning to @LinuxMercedes because wtf is this scrollback code. :ship:

### CODE REVIEW

**Comment by LinuxMercedes** [comment]
I think the best way to implement this would be to write a custom version of `getFormattedScrollbackLinesFromRanges()` that drops the numbers in front of each line. Otherwise, I think you'll end up breaking on any sort of complicated quo spec, or you'll implement your own separate language for getting stuff from scrollback, which seems like a bad idea. 

For an example of how this might work, consider:

```
01:18 < drusepth> !ezquo drusepth 3-1 suroi /kpop/ 2-1
01:18 < drubot> drusepth:
01:18 < drubot> (3) <drusepth> I'm so great
01:18 < drubot> (2) <drusepth> I built this broken feature
01:18 < drubot> (1) <drusepth> Okay time to sleep
01:18 < drubot> suroi /kpop/:
01:18 < drubot> (2) <suroi> kpop is the bomb diggity
01:18 < drubot> (1) <suroi> this is a real quote about kpop
```

(Note that the new scrollback parser lets you specify line ranges on nick, regex, and nick+regex searches!)

I am hesitant to suggest the numbering be integrated into `getFormattedScrollbackLinesFromRanges()` directly (thoughts, @euank ?). On the other hand, the `TODO`s in that function could be fixed as part of this PR since you'll be writing the code needed to implement them anyway.

Perhaps the 'best' (most general) approach would be to allow you to pass a callback to `getFormattedScrollbackLinesFromRanges()` that would get called with the line, spec, and nick/regex/line info and allow you to format the resulting line however you like. I would be happy to help implement that if that seems like a reasonable approach.

**Comment by LinuxMercedes** [comment]
From chatting with @euank : what about a `!quo?` command that works like `!quo`, except it messages you first with the text it would quote and asks whether you really want to do that. We could do that instead of `!ezquo` or in addition. 

(Implementing `!quo?` would require some architectural changes that would be nice to have anyway.)

**Comment by LinuxMercedes** [comment]
For the record,  bd7d4cd introduces some bug fixes that should definitely be merged.

File: modules/scrollback/index.js [hunk 1]
```diff
@@ -466,22 +478,26 @@ module.exports.getFormattedScrollbackLinesFromRanges = function(channel, input,
...
           else {
-            regexChunks.push(matches);
+            regexChunks.push({nick:chunk.nick, regex:regex, contents:matches});
           }
         }
       }
 
       if(regexChunks.length == 0) {
-        regexChunks.push(revCache);
+        regexChunks = nickChunks;
```
Comment (hfukada) [inline modules/scrollback/index.js]: so if no lines match the nickChunks, is it more correct to return all the nickChunks or  return []?

Comment (hfukada) [inline modules/scrollback/index.js]: still a tad confusing. talked about it

**Comment by hfukada** [comment]
sgtm 
request to rename to testquo or putsquo or printquo

**Comment by LinuxMercedes** [comment]
`!scrollback` is my vote for the name.

**Comment by drusepth** [comment]
I vote `!que`, a portmanteau of quo and query.

**Comment by LinuxMercedes** [comment]
Ok, scrollback stuff seems to be working right and ezquo has reasonable output formatting now. Someone else should pick the command name (my vote's for `que`). I'll clean up the commit history once that's all done.

@hfukada @euank LGTY?

File: modules/ezquo/index.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+var bot = null;
+
+module.exports.init = function(b) {
+  bot = b;
+};
+
+module.exports.commands = ['ezquo', 'ezquote'];
+
+module.exports.run = function(r, parts, reply, command, from, to) {
+  if (to[0] != '#' && to[0] != '&')
```
Comment (euank) [inline modules/ezquo/index.js]: nit: prefer `!bot.isChannel(to)`

**Comment by euank** [comment]
I like "que" or "quo?" as the name. This LGTM as it is, but I think long term I'd rather have something with an idea of a "staged" quo which you commit, because this won't work for fast-moving conversations etc.

**Pull request merged**